Improve Execution Strategy documentations (#13477)
eclipse/jetty.project: documentation/jetty/modules/programming-guide/pages/arch/threads.adoc

[[execution-strategy-pc]]
=== Produce-Consume

> In the `Produce-Consume` mode, the `Producer Thread` loops between producing tasks (`Pn`); and directly running those tasks (`Rn`).

Reviewer: Instead of `running those tasks (`Rn`)`, shouldn't this be `executing those tasks (`En`)` to be consistent with our existing naming (E,P,C)?

Author: Actually it should be Cn for consume.

Reviewer 2: "loops between" is confusing to me as it seems that there are two loops. Even ChatGPT says it's not idiomatic English, suggesting "loops over".

Also the semicolon is even weirder as it breaks the reading but there is nothing to break.

I liked the original better.

> In the `Produce-Consume` mode, the `Producer Thread` loops between producing tasks (`Pn`); and directly consuming (aka running) those tasks (`Cn`).

Reviewer 2: We don't use "aka", not even with dots.

```suggestion
In the `Produce-Consume` mode, the `Producer Thread` loops producing tasks (`Pn`) and directly consuming the tasks (`Cn`) by running them.

```

hide time-axis

@PT
0 is P1 #lightgreen

> 1 is C1 #dodgerblue

Reviewer 2: "C1" is less explicative of "Run T1". I liked it better before.

Author: But "Run T1" has no link to "Consume", so there is no semantic linkage from the strategy name to the explanation.  You know that running a task is consuming it, but the average reader will not.

This mode should only be used if the produced task is known to never block, or if the system tolerates well (or does not care about) head-of-line blocking.

This mode is equivalent to the "selector loop" or "event loop" models typical of non-blocking I/O frameworks such as link:https://nodejs.org/[NodeJS], link:https://netty.io/[Netty] and others.

[[execution-strategy-pec]]
=== Produce-Execute-Consume
In the `Produce-Execute-Consume` mode, the `Producer Thread` is in a production loop (`Pn`, `Pn+1`, ...) producing tasks that are submitted to a `java.util.concurrent.Executor` to be run (`Cn`, `Cn+1`, ...) by a `Worker Thread` different from the `Producer Thread`.

[plantuml]
----
skinparam backgroundColor transparent

compact concise "Producer Thread" as PT
compact concise "Worker Thread A" as WT1
compact concise "Worker Thread B" as WT2
compact concise "Worker Thread C" as WT3
hide time-axis

@PT
0 is P1 #lightgreen
1 is P2 #lightgreen
2 is P3 #lightgreen
3 is P4 #lightgreen
4 is {hidden}

@WT1

> 1 is C1 #dodgerblue

Reviewer 2: Less clear.

Author: More clear

[[execution-strategy-epc]]
=== Execute-Produce-Consume

> In the `Execute-Produce-Consume` mode, there is no dedicated producer thread. Rather a `Thread x` is initially the producer and when it has produced a task (`Pn`) it submits an internal task to an `Executor` so `Thread y` can take over production (`Pn+1`), and the original `Thread x` directly runs the task it produced (`Cn`).

Reviewer 2: Every sentence on its own line.
I find this much less clear than before, as I need to juggle between x, y, n, n+1 and other symbols that were not there before, and they are not necessary.

Before, I could glance at the diagram and understand, now it's just symbols and it's less understandable.

Author: The confusion with the previous version was precisely because everything share the same number space.  So is Thread N related to task N?  

You can still glance at the diagram?!?!?

Let me try a version that doesn't even name the threads...

[plantuml]
----
skinparam backgroundColor transparent

compact concise "Thread A" as WT1
compact concise "Thread B" as WT2
compact concise "Thread C" as WT3
compact concise "Thread D" as WT4
hide time-axis

@WT1
0 is P1 #lightgreen
1 is C1 #dodgerblue
5 is {hidden}

@WT2
1 is P2 #lightgreen
2 is C2 #dodgerblue
4 is P5 #lightgreen
5 is R5 #dodgerblue
10 is {hidden}

@WT3
2 is P3 #lightgreen
3 is C3 #dodgerblue
6 is {hidden}

@WT4
3 is P4 #lightgreen
4 is R4 #dodgerblue
8 is {hidden}

----

This mode may operate similar to <<execution-strategy-pc,`Produce-Consume`>> as after consuming a task (`Cn`) a thread may try to become a producer again (`Pn+m`), as `Thread B` does above.

> Differently from `Produce-Execute-Consume`, tasks are run by the same thread that produced it (which is CPU core cache efficient and production happens on a different thread. However, if the executor does not have an idle thread immediately available then without a thread to continue production the strategy may delay or even hang.

Reviewer 2: Missing a closing parens, and less clear.

The sentence says that "tasks are run by the same thread that produced it", and immediately "and production happens on a different thread", but you just said it was the same thread.

The original text was clearer, as it introduced the concept of production take-over that has been removed here.

> Differently from `Produce-Execute-Consume`, here production happens on different threads, but the advantage is that the task is run by the same thread that produced it (which is CPU core cache efficient). However, if the executor does not have an idle thread immediately available then using this strategy may result in production being queued waiting for an available thread, which can result in dead lock.

Reviewer: `which can result in dead lock` is too vague, or assuming too much context: there may not be any two threads mutually waiting explicitly for each other's already held lock.

I'd rather speak of a hang rather than a deadlock in this case.

* `org.eclipse.jetty.util.thread.strategy.ProduceExecuteConsume` for the `Produce-Execute-Consume` mode.
* `org.eclipse.jetty.util.thread.strategy.ExecuteProduceConsume` for the `Execute-Produce-Consume` mode.

> There is an additional implementation of `ExecutionStrategy`, namely `org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy`, that takes into account both the availability of idle threads and the `InvocationType` of the task when the task implements `Invocable` (otherwise it assumes `BLOCKING`).

Reviewer: `both the availability of idle threads and` looks like an implementation detail to me.

Author: Indeed it is, but hard to describe the adaptive nature without describing that.  How is t his variation?

> There is an additional implementation of `ExecutionStrategy`, namely `org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy`, that heuristically chooses between the other strategies. The implementation takes into account both the availability of idle threads and the `InvocationType` of the task when choosing which specific strategy to apply.

Reviewer 2: One line per sentence.

"availability of idle threads" is not correct: it does not matter if idle threads are available, there must be reserved threads available.

This concept is explained more precisely 2 paragraphs later, so I liked the original text better.

Author: The previous version was wrong as it left out half the basis of the adaption.   I've modified both this paragraph and the one 2 later so they are better linked and the later expands on this one.
